stdlib/mexpr/call-graph.mc

```diff
@@ -0,0 +1,109 @@
+-- Defines a language fragment for constructing a call-graph of the bindings in
+-- a recursive let-expression. The call graph contains an edge from a to b if
+-- the body of binding a contains a call to binding b.
+
+include "digraph.mc"
+include "mexpr/ast.mc"
+include "mexpr/ast-builder.mc"
+include "mexpr/symbolize.mc"
+include "mexpr/type-annot.mc"
+
+lang MExprCallGraph = MExprAst
+  sem constructCallGraph =
+  | TmRecLets t ->
+    let g : Digraph Name Int = digraphEmpty nameCmp eqi in
```

Comment:
What is the meaning of the `Int` labels? Unique identifiers of the labels, I guess? And why does it have to start with a `TmRecLets` node? I also have a call graph generation in my transformations, but the type of that one is `Digraph (Name, Info) (Name, Info)` (name and info of functions and call sites). I'm wondering if it's possible for us to use the same code, but I haven't looked too deeply if our use-cases are too different.

Reply:
The labels are not used, I just set the label to `0` for all edges, so they might as well have type `()`. My approach should still work with the type you used.

I implemented it for `TmRecLets` because I only need call graphs for recursive let-expressions (to find an order in which to traverse all vertices, i.e. the bindings and their local functions). If all code except that in `constructCallGraph` is compatible with your approach, then it should work, because it could just be renamed and rewritten to call a more general call graph function.